### DISCUSSION TITLE
Link between motor ratio and PWM

### DISCUSSION THREAD

**Discussion by GusBon** [bitcraze/discussions#1520] [Q&A] [answered]
Hello,

I am currently working on Bolt based UAV with the attitude firmware unchanged (tuned, but no structure changes) and I am trying to build a model of my system.

To do so I am currently doing system identification of thrust and torque coefficient of my motor+propeller. I plan to use the Tyto Robotics 1585 thrust stand (https://www.tytorobotics.com/pages/series-1580-1585). With this thrust stand I would be able to send PWM ratios and measure RPM, thrust and torque of my thruster (motor+propeller). With those measures I would be able to identify the missing coefficients.

My question is:
What is the link between the motor ratio value (which is a ratio in the form of an INT16) and the actual PWM sent to the ESC ? 
I know that Bolt based UAV seems to use the DSHOT protocol, but my question is more like: if the value of motor ratio is equivalent to 50% (INT16_max/2) then do I get the same behavior (same RPM) when I am setting a PWM at 50% on my thrust stand test bench ?

If yes, I could then convert motor ratio to to RPM and then thrust and torque produced by my thruster in my model.

If no, then do someone has a sort of conversion formula to do so ?

(If someone is interested by the result of system identification I could post the results here (if the process is successful), the UAV is equipped with Gemfan 65mm -1.5mm propeller and Nin v2 1203pro-5500KV Flywoo motors).

Thanks in advance ! 
Gus

**Comment by tobbeanton** [comment, marked as answer]
Hi Gus,

For brushless motors the "battery voltage compensation function" in not active so the output is fully linear, INT16_max/2 -> 50% motor signal output.

Please share the result when you have it!

**Reply by GusBon** [reply]
Hi,

First thanks a lot for your answer !

I was able to perform tests with the test bench, here are some results. They might not be the more accurate results but I have done the best I could with the constraints I had.

First the Nin v2 motor is too small for the test bench, so I had to build an adapter.

Then I obtained those coefficients:
$ratio2rpmcoeff= 4.989\times10^3$
$ratio2rpmbias = 327,7464$
$thrustcoeff = 1.564\times10^{-8}$
$torquecoeff = 8.845\times10^{-11}$

As:
$rpm = ratio2rpmcoeff\times ratio + ratio2rpmbias$
The link between the RPM of the motor (in rad/s) and the input ratio signal (value between 0 and 1 considering that the minimum signal sent by the 50 Hz PWM is 1000us and the maximum signal sent is 2000us). Therefore this equation represents the link between the variable motor ratio (normalized) and actual RPM of the thruster.

$thrust = thrustcoeff\times{\omega^2}$
The link between the thrust produced by a thruster rotating at $\omega$ rad/s.

$torque = torquecoeff\times{\omega^2}$
The link between the drag torque produced by a thruster rotating at $\omega$ rad/s.

Here is some comparison graphs between measurements and estimation:
[ratio_vs_rpm_comparison.pdf](https://github.com/user-attachments/files/16677469/ratio_vs_rpm_comparison.pdf)
[rpm_vs_thrust_comparison.pdf](https://github.com/user-attachments/files/16677470/rpm_vs_thrust_comparison.pdf)
[rpm_vs_torque_comparison.pdf](https://github.com/user-attachments/files/16677471/rpm_vs_torque_comparison.pdf)

As you can see results seem pretty accurate for RPM conversion and thrust, but for the torque coefficient does not look to fit regarding our measurements. Off the top of my head it might be because our motor is small compared to the used thrust stand.

**To conclude this data is meant to be taken with caution as the measured thruster does not look to be adapted for the thrust stand!**

**Reply by tobbeanton** [reply]
Thanks for this! We rebuilt the thrust stand to be more sensitive. From [this](https://www.bitcraze.io/2023/05/modifying-a-thrust-stand-for-small-motors/) blog post you can get some inspiration.

**Reply by GusBon** [reply]
Hi !

This seems to be a nice workaround to test small thrusters, thanks for the hint ! Unfortunately, I would not be able to redo measurements, but I will clearly save this blog post for the future.
